plugin/src/main/java/org/opensearch/ml/action/models/DeleteModelTransportAction.java

```diff
@@ -64,4 +77,35 @@ public void onFailure(Exception e) {
         }
     }
 
+    @VisibleForTesting
+    void deleteModelChunks(String modelId, DeleteResponse deleteResponse, ActionListener<DeleteResponse> actionListener) {
+        DeleteByQueryRequest deleteModelsRequest = new DeleteByQueryRequest(ML_MODEL_INDEX);
+        deleteModelsRequest.setQuery(new TermsQueryBuilder(MODEL_ID_FIELD, modelId));
+
+        client.execute(DeleteByQueryAction.INSTANCE, deleteModelsRequest, ActionListener.wrap(r -> {
+            if ((r.getBulkFailures() == null || r.getBulkFailures().size() == 0)
+                && (r.getSearchFailures() == null || r.getSearchFailures().size() == 0)) {
+                log.info("All model chunks are deleted for model {}", modelId);
+                actionListener.onResponse(deleteResponse);
+            } else {
+                logFailure(r, modelId, actionListener);
+            }
+        }, e -> {
+            log.info("Failed to delete ML model for " + modelId, e);
+            actionListener.onFailure(e);
+        }));
+    }
+
+    private void logFailure(BulkByScrollResponse response, String modelId, ActionListener<DeleteResponse> actionListener) {
+        String errorMessage = "";
+        if (response.isTimedOut()) {
+            errorMessage = OS_STATUS_EXCEPTION_MESSAGE + "," + TIMEOUT_MSG + modelId;
+        } else if (!response.getBulkFailures().isEmpty()) {
+            errorMessage = OS_STATUS_EXCEPTION_MESSAGE + "," + BULK_FAILURE_MSG + modelId;
+        } else {
+            errorMessage = OS_STATUS_EXCEPTION_MESSAGE + "," + SEARCH_FAILURE_MSG + modelId;
+        }
+        log.info(response.toString());
+        actionListener.onFailure(new OpenSearchStatusException(errorMessage, RestStatus.INTERNAL_SERVER_ERROR));
```

Comment:
What does `Bulk failure` and `search failure` mean?

 If search failure mean: customer is provided a model id and we didn't find that model to delete, then I don't think that error should be treated as `Internal server error 500`.
 
Just curious, if customers see `Bulk failure` what do they suppose to do?

Reply:
Check this doc https://www.javadoc.io/doc/org.elasticsearch/elasticsearch/6.0.1/org/elasticsearch/index/reindex/ScrollableHitSource.SearchFailure.html. If no model is found in search, it just doesn't delete and wouldn't return an error. It only delete model chunks that are returned in search results. 

Bulk failure means a list of Failures, Users can use the search API to see which chunks are deleted. If the model metadata is deleted, a cleanup job will remove the rest of the chunks.